File: app/Models/Order.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasMany;

class Order extends Model
{
    protected $fillable = [
        'order_number', 'user_id', 'customer_name', 'customer_phone', 'customer_email',
        'city', 'area', 'address_line', 'building', 'delivery_notes', 'delivery_zone_id',
        'subtotal', 'delivery_fee', 'discount_amount', 'total',
        'coupon_code', 'coupon_id', 'status', 'payment_method', 'payment_status',
        'admin_notes', 'confirmed_at', 'shipped_at', 'delivered_at',
        'last_edited_by', 'last_edited_at',
        'loyalty_points_redeemed', 'loyalty_discount',
        'return_requested_at', 'return_reason', 'return_status', 'return_reject_reason',
    ];

    protected $casts = [
        'subtotal'        => 'decimal:2',
        'delivery_fee'    => 'decimal:2',
        'discount_amount' => 'decimal:2',
        'total'           => 'decimal:2',
        'confirmed_at'         => 'datetime',
        'shipped_at'           => 'datetime',
        'delivered_at'         => 'datetime',
        'last_edited_at'       => 'datetime',
        'return_requested_at'  => 'datetime',
    ];

    public static $statusLabels = [
        'pending'    => 'بانتظار التأكيد',
        'confirmed'  => 'تم التأكيد',
        'processing' => 'قيد التجهيز',
        'shipped'    => 'في الطريق',
        'delivered'  => 'تم التوصيل',
        'cancelled'  => 'ملغي',
        'returned'   => 'مُرجَّع',
    ];

    public static $statusColors = [
        'pending'    => 'warning',
        'confirmed'  => 'info',
        'processing' => 'info',
        'shipped'    => 'primary',
        'delivered'  => 'success',
        'cancelled'  => 'danger',
        'returned'   => 'danger',
    ];

    protected static function boot(): void
    {
        parent::boot();

        static::creating(function ($order) {
            if (empty($order->order_number)) {
                $order->order_number = 'ORD-' . date('Y') . '-' . str_pad(
                    (Order::whereYear('created_at', date('Y'))->count() + 1),
                    4, '0', STR_PAD_LEFT
                );
            }
        });

        // ── Audit: log inline status changes (and any other model updates) ──
        static::updating(function (Order $order) {
            if (! auth()->check()) return;

            $original = $order->getOriginal();
            $changes  = [];
            foreach ($order->getDirty() as $field => $newVal) {
                // skip noise + skip fields that EditOrder page already logs explicitly
                if (in_array($field, ['updated_at', 'last_edited_at', 'last_edited_by'])) continue;
                $changes[$field] = ['before' => $original[$field] ?? null, 'after' => $newVal];
            }
            if (empty($changes)) return;

            // Skip if this update is from the EditOrder page (it sets last_edited_at)
            if ($order->isDirty('last_edited_at')) return;

            $action = (count($changes) === 1 && isset($changes['status'])) ? 'status_change' : 'edit';

            // Defer creation of audit until AFTER the update finished
            $order->_pendingAudit = ['action' => $action, 'changes' => $changes];
        });

        static::updated(function (Order $order) {
            if (! empty($order->_pendingAudit)) {
                OrderAudit::log($order, $order->_pendingAudit['action'], $order->_pendingAudit['changes']);
                $order->_pendingAudit = null;
            }

            // ── Award loyalty points when order reaches the configured status ──
            $earnStatus = (string) \App\Models\Setting::get('earn_on_status', 'delivered');
            if (
                $order->wasChanged('status')
                && $order->status === $earnStatus
                && \App\Services\LoyaltyService::enabled()
                && $order->user_id
            ) {
                \App\Services\LoyaltyService::awardForOrder($order);
            }

            // ── Push notification to customer on status change ──
            if ($order->wasChanged('status') && $order->user_id) {
                $customer = User::find($order->user_id);
                if ($customer && $customer->fcm_token) {
                    $statusLabel = self::$statusLabels[$order->status] ?? $order->status;
                    \App\Services\FcmService::sendToUser(
                        $customer,
                        '📦 تحديث على طلبك',
                        "طلب #{$order->order_number} • الحالة: {$statusLabel}",
                        [
                            'order_number' => $order->order_number,
                            'status'       => $order->status,
                            'type'         => 'order_status_change',
                        ]
                    );
                }
            }
        });
    }

    /** Transient property to defer audit logging until after update. */
    public ?array $_pendingAudit = null;

    public function user(): BelongsTo         { return $this->belongsTo(User::class); }
    public function deliveryZone(): BelongsTo { return $this->belongsTo(DeliveryZone::class); }
    public function coupon(): BelongsTo       { return $this->belongsTo(Coupon::class); }
    public function items(): HasMany          { return $this->hasMany(OrderItem::class); }
    public function audits(): HasMany         { return $this->hasMany(OrderAudit::class)->latest(); }
    public function lastEditor(): BelongsTo   { return $this->belongsTo(User::class, 'last_edited_by'); }

    public function getStatusLabelAttribute(): string
    {
        return self::$statusLabels[$this->status] ?? $this->status;
    }

    public function getWhatsappMessageAttribute(): string
    {
        $msg = "مرحباً {$this->customer_name} 👋\n\n";
        $msg .= "تم استلام طلبك بنجاح! 🎉\n";
        $msg .= "رقم الطلب: #{$this->order_number}\n";
        $msg .= "الإجمالي: {$this->total} ₪\n\n";
        $msg .= "سنتواصل معك قريباً لتأكيد موعد التوصيل.\n";
        $msg .= "شكراً لثقتك بأبناء الفريد 🛍️";
        return urlencode($msg);
    }
}

File: mobile-app/lib/models/order.dart
import '../utils/image_helper.dart';

class AppOrder {
  final String orderNumber;
  final String status;
  final String statusLabel;
  final double subtotal;
  final double deliveryFee;
  final double discountAmount;
  final double loyaltyDiscount;
  final double total;
  final DateTime? createdAt;
  final String? createdAtHuman;
  final int itemsCount;

  // Detail-only
  final String? customerName;
  final String? customerPhone;
  final String? city;
  final String? area;
  final String? addressLine;
  final String? building;
  final String? paymentMethod;
  final List<OrderItem> items;
  final DateTime? returnRequestedAt;

  AppOrder({
    required this.orderNumber,
    required this.status,
    required this.statusLabel,
    required this.subtotal,
    required this.deliveryFee,
    required this.discountAmount,
    required this.loyaltyDiscount,
    required this.total,
    this.createdAt,
    this.createdAtHuman,
    this.itemsCount = 0,
    this.customerName,
    this.customerPhone,
    this.city,
    this.area,
    this.addressLine,
    this.building,
    this.paymentMethod,
    this.items = const [],
    this.returnRequestedAt,
  });

  factory AppOrder.fromJson(Map<String, dynamic> j) => AppOrder(
    orderNumber:    j['order_number']    as String? ?? '',
    status:         j['status']          as String? ?? 'pending',
    statusLabel:    j['status_label']    as String? ?? '',
    subtotal:      (j['subtotal']        as num?)?.toDouble() ?? 0,
    deliveryFee:   (j['delivery_fee']    as num?)?.toDouble() ?? 0,
    discountAmount:(j['discount_amount'] as num?)?.toDouble() ?? 0,
    loyaltyDiscount:(j['loyalty_discount'] as num?)?.toDouble() ?? 0,
    total:         (j['total']           as num?)?.toDouble() ?? 0,
    createdAt:      j['created_at'] != null ? DateTime.tryParse(j['created_at']) : null,
    createdAtHuman: j['created_at_human'] as String?,
    itemsCount:    (j['items_count']    as num?)?.toInt() ?? 0,
    customerName:   j['customer_name']  as String?,
    customerPhone:  j['customer_phone'] as String?,
    city:           j['city']           as String?,
    area:           j['area']           as String?,
    addressLine:    j['address_line']   as String?,
    building:       j['building']       as String?,
    paymentMethod:      j['payment_method']       as String?,
    returnRequestedAt:  j['return_requested_at'] != null
        ? DateTime.tryParse(j['return_requested_at']) : null,
    items: ((j['items'] as List?) ?? const [])
        .map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
        .toList(),
  );
}

class OrderItem {
  final int id;
  final int productId;
  final String productName;
  final String? productImage;
  final double price;
  final int quantity;
  final double total;
  final String? variantInfo;

  OrderItem({
    required this.id,
    required this.productId,
    required this.productName,
    this.productImage,
    required this.price,
    required this.quantity,
    required this.total,
    this.variantInfo,
  });

  factory OrderItem.fromJson(Map<String, dynamic> j) => OrderItem(
    id:           _toInt(j['id']),
    productId:    _toInt(j['product_id']),
    productName:  j['product_name'] as String? ?? '',
    productImage: ImageHelper.cleanUrl(j['product_image'] as String?),
    price:       (j['price']        as num?)?.toDouble() ?? (double.tryParse(j['price'].toString()) ?? 0),
    quantity:    (j['quantity']     as num?)?.toInt() ?? 1,
    total:       (j['total']        as num?)?.toDouble() ?? (double.tryParse(j['total'].toString()) ?? 0),
    variantInfo:  j['variant_info'] as String?,
  );

  static int _toInt(dynamic v) {
    if (v is int) return v;
    if (v is num) return v.toInt();
    return int.tryParse(v?.toString() ?? '0') ?? 0;
  }
}
